{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "# sys.path.append('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## test 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "<>:493: SyntaxWarning: invalid escape sequence '\\o'\n",
      "<>:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "<>:493: SyntaxWarning: invalid escape sequence '\\o'\n",
      "/tmp/ipykernel_12253/1907232243.py:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "  pattern=\">fig.(\\d+\\.\\d+.peg.\\d+) (.+)\"\n",
      "/tmp/ipykernel_12253/1907232243.py:493: SyntaxWarning: invalid escape sequence '\\o'\n",
      "  '''\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Cluster Analysis Module\n",
    "-----------------------\n",
    "Responsible for extracting the cluster attributes from the pipeline output and setting them as node attributes in the networkx graph\n",
    "\n",
    "The pipeline output consists of 4 files that we need:\n",
    "\n",
    "    - *.fatsa.clstr file (cd hit output) that contains the cluster number and the sequences in all of the genomes\n",
    "    - *_cluster_frequencies.csv file that contains the cluster number and the frequency of the cluster in the genomes\n",
    "    - *_pangenome.csv file that contains the cluster number and the gene class of the cluster\n",
    "    - *.fasta file (cd hit output) that contains the cluster number and the representative sequence of the cluster4\n",
    "\n",
    "Attributes:\n",
    "----------\n",
    "    _output_path: one global (private) attribute that is the path to the pipeline output directory - changed using a setter `set_output_path(str)`\n",
    "\n",
    "Functions: \n",
    "----------\n",
    " \n",
    "note: even though there are many helper functions, kept public in case needed (so far)\n",
    "\n",
    "* generate_files_paths(pipeline_output_path:str)->str, str, str, str\n",
    "* get_cluster_frequency(freq_path:str)->pd.DataFrame\n",
    "* get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame\n",
    "* get_representative_products(clstr_fasta_path:str)->pd.DataFrame\n",
    "* combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame\n",
    "* get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame\n",
    "* get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame\n",
    "* get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path)\n",
    "* create_dicts(df:pd.DataFrame)->dict\n",
    "* set_node_attributes_by_cluster(df:pd.DataFrame, G:nx.graph)->nx.Graph\n",
    "* set_output_path(path:str)->str\n",
    "* set_cluster_attributes(G:nx.Graph, pipeline_output=_output_path)->nx.Graph\n",
    "* split_matrix_by_phenotype(unlabeled_presence_df:pd.DataFrame, pheno_df: pd.DataFrame)->(pd.DataFrame, pd.DataFrame)\n",
    "* generate_RS_presence_counts(R:pd.DataFrame, S:pd.DataFrame)->pd.DataFrame\n",
    "* compute_log_odds_ratio(RS_counts_df:pd.DataFrame)->pd.DataFrame\n",
    "* get_cluster_resistance_LOR(presence_df:pd.DataFrame, pheno_df:pd.DataFrame)-> pd.DataFrame\n",
    "    \n",
    "'''\n",
    "\n",
    "\n",
    "import requests\n",
    "import re\n",
    "import os\n",
    "import sys\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def generate_files_paths(pipeline_output_path):\n",
    "    '''\n",
    "    takes the output dir path of the pipeline for a species, and returns the 4 file paths inside of it needed to extract the data\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    pipeline_output_path: str, path to the pipeline output directory\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    clstr_path: str, path to the cluster file, cdhit output\n",
    "    freq_path: str, path to the cluster frequency file\n",
    "    pan_annot_path: str, path to the pangenome annotation file\n",
    "    clstr_fasta_path: str, path to the cluster fasta file, aslo cd hit output\n",
    "    '''\n",
    "\n",
    "    pipeline_output_path=pipeline_output_path.rstrip('/')\n",
    "    basename=os.path.basename(pipeline_output_path)\n",
    "\n",
    "    clstr_path=pipeline_output_path+'/'+basename+'.fasta.clstr'\n",
    "    freq_path=pipeline_output_path+'/'+basename+'_cluster_frequencies.csv'\n",
    "    pan_annot_path=pipeline_output_path+'/'+basename+'_pangenome.csv'\n",
    "    clstr_fasta_path=pipeline_output_path+'/'+basename+'.fasta'\n",
    "\n",
    "    return clstr_path, freq_path, pan_annot_path, clstr_fasta_path\n",
    "\n",
    "# clstr_path, freq_path, pan_annot_path, clstr_fasta_path= generate_files_paths(\"../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/\")\n",
    "\n",
    "def get_cluster_frequency(freq_path):\n",
    "    '''\n",
    "    Takes the path to the cluster frequency file and returns a dataframe with the cluster frequencies\n",
    "\n",
    "    Parameters:\n",
    "    freq_path: str\n",
    "\n",
    "    Returns:\n",
    "    freq_df: pandas.DataFrame\n",
    "    '''\n",
    "\n",
    "    freq_df=pd.read_csv(freq_path, index_col=0)\n",
    "    freq_df.columns=['Cluster','frequency']\n",
    "    freq_df.index=freq_df['Cluster']\n",
    "    freq_df.drop('Cluster', axis=1, inplace=True)\n",
    "\n",
    "    return freq_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_representative_products(clstr_fasta_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a CD-HIT fasta output and returns a dataframe that matches each CDS PATRIC ID with its product name\n",
    "\n",
    "    the fasta output is of this form:\n",
    "\n",
    "    >fig|195.2024.peg.83 Putative oxidoreductase ferredoxin-type protein, clusters with CPO\n",
    "    MNFSQISDACVKCGKCIPVCTIHEVNRDETTSPRGFLDLLAAYKEEKLELDKEAKKIFES\n",
    "    CFLCTNCVEVCPSKLRVDNVIEEVRYDIAKKFGIAWYKKIIFFFLRRRKILDLVAKLGYV\n",
    "    FQSCAFKIQSQDQNVGMKAKFSMPFVKKGRLLTSFNKKSFLNSNPDFIDNGGEKTVGFFV\n",
    "    GCLANYFYIDTANAVLKIAKEVKINVDLMKEQVCCGAPQFFTGDFKSVEILAKKNIEYFE\n",
    "    KKLEKLDAIIIPEATCSAMLKIDLEHFFNMQNEPEWAKRAQKISSRIYMASEYFYKFTNL\n",
    "    KELLESKKKLNYSITYHDPCHARKMQGVFKEPRELLKANYHFVEMSNPNACCGFGGVSMQ\n",
    "    TDYYDRALSVGLKKASMIDESKACVVSAECSACRMQISNALEQNSSKAIFASPLELIAKA\n",
    "    L\n",
    "    >fig|195.2024.peg.542 Septum-associated rare lipoprotein A\n",
    "    MKPYTINGKTYYPTVVSVGETADGIASWYGPGFHGKKTSNGETYNQNGLTAAHKTLPMNT\n",
    "    ILKVTNLNNNRQVTVRVNDRGPFVNNRIIDLSKGAASQIDMIAAGTAPVRLEVIGFGSAN\n",
    "    SGNNVVHSNINYGASGGIANNGQIYEGGNFMVQIGAFKNPSGAQTIASRYKTYRTYSSTI\n",
    "    RKSSVDGLSRVFLTGFRSEEEARDFAASGAFAGAFVVRE\n",
    "\n",
    "    The aim is to match 195.2024.peg.542 with Septum-associated rare lipoprotein A in teh dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_fasta_path: str, path of the fasta output\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, df of columns gene_representative and product_name\n",
    "    '''\n",
    "\n",
    "    dict={}\n",
    "\n",
    "    with open(clstr_fasta_path,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith(\">\"):\n",
    "                pattern=\">fig.(\\d+\\.\\d+.peg.\\d+) (.+)\"\n",
    "                gene_representative=re.match(string=line, pattern=pattern).group(1)\n",
    "                product_name=re.match(string=line, pattern=pattern).group(2)\n",
    "                dict[gene_representative]=product_name\n",
    "    \n",
    "\n",
    "    df = pd.DataFrame(dict.items(), columns=['gene_representative', 'product_name'])\n",
    "    \n",
    "    return df\n",
    "\n",
    "def combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Combines the cluster representative dataframe with the product name dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_rep_df: pd.DataFrame, dataframe with cluster number and gene representative (get_cluster_representatives() output)\n",
    "    - rep_prod_df: pd.DataFrame, dataframe with gene representative and product name (get_representative_products() output)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, dataframe with cluster number, gene representative and product name\n",
    "    '''\n",
    "    # if clstr rep df doesnt have a column named Cluster, create one out of index\n",
    "    if 'Cluster' not in clstr_rep_df.columns:\n",
    "        clstr_rep_df['Cluster']=clstr_rep_df.index\n",
    "\n",
    "    df=pd.merge(clstr_rep_df, rep_prod_df, on='gene_representative', how='left')\n",
    "    df.drop(columns='gene_representative', inplace=True)\n",
    "\n",
    "    df.set_index('Cluster', inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a pangenome annotation file and returns a dataframe with cluster number and gene class\n",
    "    This annotation file is a <species>_pangenome.csv file output from the pangenome analysis pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - pan_annot_path: str, path to the pangenome annotation file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - pan_df: pd.DataFrame, dataframe with cluster number and gene class\n",
    "    '''\n",
    "    with open(pan_annot_path, 'r') as f:\n",
    "        pan_df=pd.read_csv(f, index_col=0)\n",
    "        pan_df.drop(pan_df.columns[[1]], axis=1, inplace=True)\n",
    "        pan_df.columns=['Cluster', 'pan_gene_class']\n",
    "\n",
    "    pan_df.set_index('Cluster', inplace=True)\n",
    "    return pan_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path):\n",
    "\n",
    "    clstr_patric_id_df=get_cluster_representatives(clstr_path)\n",
    "    clstr_gene_class_df= get_cluster_pan_gene_class(pan_annot_path)\n",
    "    _patric_id_product_df=get_representative_products(clstr_fasta_path)\n",
    "    clstr_product_df=combine_cluster_product(clstr_patric_id_df, _patric_id_product_df)\n",
    "    clstr_freq_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "    df=pd.concat([clstr_patric_id_df, clstr_gene_class_df, clstr_product_df, clstr_freq_df], axis=1)\n",
    "\n",
    "    #if there exist a Cluster col remove\n",
    "    if 'Cluster' in df.columns:\n",
    "        df.drop('Cluster', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def create_dicts(df:pd.DataFrame):\n",
    "    '''\n",
    "    takes a nxm data frame and returns m dictionaries of keys the index, and values each col repectively\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - dicts: dict, dict of m dictionaries - key is the col name, value is the dictionary of the col\n",
    "    '''\n",
    "    dicts = {col: df[col].to_dict() for col in df.columns}\n",
    "    return dicts\n",
    "\n",
    "def set_node_attributes_by_cluster(df:pd.DataFrame, G:nx.graph):\n",
    "    '''\n",
    "    takes a dataframe and a graph and sets the node attributes for each cluster\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "    - G: nx.Graph, graph to set the node attributes\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - G: nx.Graph, graph with node attributes set\n",
    "    '''\n",
    "    dicts = create_dicts(df)\n",
    "\n",
    "    for key in dicts.keys():\n",
    "        nx.set_node_attributes(G, dicts[key], key)\n",
    "\n",
    "    return G\n",
    "\n",
    "_output_path='../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/'\n",
    "\n",
    "def set_output_path(path:str):\n",
    "    '''\n",
    "    sets the output path for the pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - path: str, path to the output directory\n",
    "    '''\n",
    "    global _output_path\n",
    "    _output_path=path\n",
    "    return _output_path\n",
    "\n",
    "def set_cluster_attributes(G:nx.Graph, pipeline_output=_output_path)->nx.Graph:\n",
    "    '''\n",
    "    takes a Graph and the pipeline's output path and performs all of the following: \n",
    "\n",
    "    - get all the files paths required for the attributes\n",
    "    - generate df for each and concatenate them all in one of index Cluster\n",
    "    - create dictionaries out of it and perform nx.set_node_attributes using each on G\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    G: nx.Graph, network of nodes Clusters\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    G: nx.Graph, network of nodes Clusters, attributes: pan_gene_class, gene_representative, pan_gene_class, frequency\n",
    "    '''\n",
    "\n",
    "    p1, p2, p3, p4 = generate_files_paths(pipeline_output)\n",
    "    df = get_cluster_attributes(p1, p2, p3, p4)\n",
    "    G = set_node_attributes_by_cluster(df, G)\n",
    "\n",
    "    return G\n",
    "\n",
    "# --------------- LOR calculation ----------------\n",
    "\n",
    "def generate_phenotype_df(pheno_path:str, presence_df:pd.DataFrame=None)->pd.DataFrame:\n",
    "    '''\n",
    "    takes in a path to a phenotype file to make a df out of it\n",
    "    if a presence_df is provided, it will filter the pheno_df to only include samples that are in the presence_df\n",
    "\n",
    "    param:\n",
    "    ----\n",
    "        - pheno_path: str\n",
    "        - presence_df: pd.DataFrame (optional)\n",
    "\n",
    "    return:\n",
    "    ----\n",
    "        - pheno_df: pd.DataFrame\n",
    "    '''\n",
    "    pheno_df = pd.read_csv(pheno_path, index_col=0)\n",
    "\n",
    "    #samples stripped\n",
    "    if presence_df is not None:\n",
    "        samples_presence= presence_df.columns\n",
    "        mask = pheno_df.index.astype(str).isin(samples_presence)\n",
    "        pheno_df = pheno_df.loc[mask]\n",
    "\n",
    "    return pheno_df\n",
    "\n",
    "def split_matrix_by_phenotype(unlabeled_presence_df:pd.DataFrame, pheno_df: pd.DataFrame)->(pd.DataFrame, pd.DataFrame):\n",
    "    '''\n",
    "    takes a gene absence presence dataframe (n x m) with the samples classification and splits it to 2 dataframes:\n",
    "        - one that represents all R samples (samples found in R_list)\n",
    "        - one that represents all S samples (found in S_list)\n",
    "    and returns them in this order: R then S\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "        - unlabeled_presence_df: (pd.DataFrame) the data frame (GxS)\n",
    "        - R_list: (list) samples genome_id that have a R (1) phenotype\n",
    "        - S_list: (list) samples having the 0 phenotype\n",
    "    \n",
    "    return:\n",
    "    ------\n",
    "        - R_df: (pd.DataFrame) data frame where columns are only for R samples\n",
    "        - S_df: (pd.DataFrame) df where cols are only for S samples\n",
    "\n",
    "    '''\n",
    "\n",
    "\n",
    "    # make sure pheno_df doesnthave more samples than those specified in the presence_df, subsample it otherwise\n",
    "    samples_presence= unlabeled_presence_df.columns\n",
    "    mask = pheno_df.index.astype(str).isin(samples_presence)\n",
    "    pheno_df = pheno_df.loc[mask]\n",
    "\n",
    "    # get the list of R and S samples:\n",
    "    R=[];S=[];U=[]\n",
    "    for sample in pheno_df.index:\n",
    "\n",
    "        if pheno_df.loc[sample].values[0]==1:\n",
    "            sample=str(sample)\n",
    "            R.append(str(sample))\n",
    "        elif pheno_df.loc[sample].values[0]==0:\n",
    "            sample=str(sample)\n",
    "            S.append(str(sample))\n",
    "        else:\n",
    "            sample=str(sample)\n",
    "            U.append(str(sample))\n",
    "\n",
    "    # All we care for care R and S which designate the list of resistant and susceptible samples respectively\n",
    "\n",
    "    R_df = unlabeled_presence_df[R]\n",
    "    S_df = unlabeled_presence_df[S]\n",
    "\n",
    "    #make a subdf of all columns that are in S of the presence_df\n",
    "\n",
    "\n",
    "    return R_df, S_df\n",
    "\n",
    "def generate_RS_presence_counts(R:pd.DataFrame, S:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes the presence matrix of R samples and S samples and returns a dataframe of the count of genes present and absent in each group of samples.\n",
    "\n",
    "    *This will memoize the entries needed for the log odds ratio computation.*\n",
    "\n",
    "    \n",
    "    e.g., output (w\\o the log odds):\n",
    "\n",
    "    | Gene       | R_present | R_absent | S_present | S_absent |\n",
    "    |------------|-----------|----------|-----------|----------|\n",
    "    | group_1001 | 96        | 0        | 187       | 0        |\n",
    "    | tig        | 96        | 0        | 187       | 0        |\n",
    "    | legF_1     | 96        | 0        | 187       | 0        |\n",
    "\n",
    "    param:\n",
    "    ----------------\n",
    "        - R: (pd.DataFrame) presence matrix of R samples (output of get_subdf_cols)\n",
    "        - S: (pd.DataFrame) presence matrix of S samples\n",
    "\n",
    "    return:\n",
    "    ----------------\n",
    "        - new_df: (pd.DataFrame) dataframe of the count of genes present and absent in each group of samples.\n",
    "\n",
    "    NOTE: it will perform the 0.5 correction - If for one gene any of the counts is 0, it will add 0.5 to all counts for that gene.\n",
    "\n",
    "    '''\n",
    "    R_present=R.sum(axis=1)\n",
    "    R_absent=R.shape[1] - R_present\n",
    "    S_present=S.sum(axis=1)\n",
    "    S_absent=S.shape[1]-S_present\n",
    "\n",
    "    indices=R.index\n",
    "\n",
    "    new_df=pd.DataFrame({'R_present':R_present, 'R_absent':R_absent, 'S_present':S_present, 'S_absent':S_absent})\n",
    "    new_df.index=indices\n",
    "\n",
    "    #the 0.5 correction:\n",
    "    for index in new_df.index:\n",
    "        row=list(new_df.loc[index])\n",
    "        if 0 in row: #check if any of the cols have value 0\n",
    "            new_df.loc[index]=new_df.loc[index]+0.5\n",
    "\n",
    "    return new_df\n",
    "\n",
    "def compute_log_odds_ratio(RS_counts_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    takes the RS_counts_df and computes the log odds ratio for each gene in the dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "        - RS_counts_df: pd.DataFrame\n",
    "\n",
    "\n",
    "    return:\n",
    "    ------\n",
    "        - log_odds_df: pd.DataFrame\n",
    "    '''\n",
    "\n",
    "    # n_R=len(R); n_S=len(S)\n",
    "\n",
    "\n",
    "    R_present=RS_counts_df['R_present']\n",
    "    R_absent=RS_counts_df['R_absent']\n",
    "    S_present=RS_counts_df['S_present']\n",
    "    S_absent=RS_counts_df['S_absent']\n",
    "\n",
    "    log_odds=np.log((R_present/R_absent)/(S_present/S_absent))\n",
    "\n",
    "    # print(log_odds)\n",
    "\n",
    "    df=pd.DataFrame({\"log_odds\":log_odds})\n",
    "    return df\n",
    "\n",
    "def get_cluster_resistance_LOR(presence_df:pd.DataFrame, pheno_path:str)-> pd.DataFrame:\n",
    "    '''\n",
    "    Knits all the steps from splitting the matrix to computing the log odds ratio for each gene in the matrix\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "        * presence_df: pd.DataFrame, GxS\n",
    "        * pheno_df: str, path to the phenotype file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "        * log_odds_df: pd.DataFrame, Gx1\n",
    "    '''\n",
    "    pheno_df = generate_phenotype_df(pheno_path, presence_df)\n",
    "    \n",
    "    R,S=split_matrix_by_phenotype(presence_df, pheno_df)\n",
    "    RS_counts = generate_RS_presence_counts(R, S)\n",
    "    log_odds_df = compute_log_odds_ratio(RS_counts)\n",
    "\n",
    "    return log_odds_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R df of  (8021, 83)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rayane/miniconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: invalid value encountered in log\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "pheno = '../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv'\n",
    "presence=pd.read_csv('../../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv', index_col=0 )\n",
    "\n",
    "lor_df= get_cluster_resistance_LOR(pheno_path=pheno, presence_df=presence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log_odds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>1.270365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>-1.220273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>0.104140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              log_odds\n",
       "Cluster 0     1.270365\n",
       "Cluster 1    -1.220273\n",
       "Cluster 2     0.104140\n",
       "Cluster 3    -0.311244\n",
       "Cluster 4     0.311244\n",
       "...                ...\n",
       "Cluster 8016 -0.311244\n",
       "Cluster 8017 -0.311244\n",
       "Cluster 8018 -0.311244\n",
       "Cluster 8019 -0.311244\n",
       "Cluster 8020 -0.311244\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lor_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# test 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "<>:503: SyntaxWarning: invalid escape sequence '\\o'\n",
      "<>:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "<>:503: SyntaxWarning: invalid escape sequence '\\o'\n",
      "/tmp/ipykernel_29791/3751185155.py:203: SyntaxWarning: invalid escape sequence '\\d'\n",
      "  pattern=\">fig.(\\d+\\.\\d+.peg.\\d+) (.+)\"\n",
      "/tmp/ipykernel_29791/3751185155.py:503: SyntaxWarning: invalid escape sequence '\\o'\n",
      "  '''\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Cluster Analysis Module\n",
    "-----------------------\n",
    "Responsible for extracting the cluster attributes from the pipeline output and setting them as node attributes in the networkx graph\n",
    "\n",
    "The pipeline output consists of 4 files that we need:\n",
    "\n",
    "    - *.fatsa.clstr file (cd hit output) that contains the cluster number and the sequences in all of the genomes\n",
    "    - *_cluster_frequencies.csv file that contains the cluster number and the frequency of the cluster in the genomes\n",
    "    - *_pangenome.csv file that contains the cluster number and the gene class of the cluster\n",
    "    - *.fasta file (cd hit output) that contains the cluster number and the representative sequence of the cluster4\n",
    "\n",
    "Attributes:\n",
    "----------\n",
    "    _output_path: one global (private) attribute that is the path to the pipeline output directory - changed using a setter `set_output_path(str)`\n",
    "\n",
    "Functions: \n",
    "----------\n",
    " \n",
    "note: even though there are many helper functions, kept public in case needed (so far)\n",
    "\n",
    "* generate_files_paths(pipeline_output_path:str)->str, str, str, str\n",
    "* get_cluster_frequency(freq_path:str)->pd.DataFrame\n",
    "* get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame\n",
    "* get_representative_products(clstr_fasta_path:str)->pd.DataFrame\n",
    "* combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame\n",
    "* get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame\n",
    "* get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame\n",
    "* get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path)\n",
    "* create_dicts(df:pd.DataFrame)->dict\n",
    "* set_node_attributes_by_cluster(df:pd.DataFrame, G:nx.graph)->nx.Graph\n",
    "* set_output_path(path:str)->str\n",
    "* set_cluster_attributes(G:nx.Graph, pipeline_output=_output_path)->nx.Graph\n",
    "* split_matrix_by_phenotype(unlabeled_presence_df:pd.DataFrame, pheno_df: pd.DataFrame)->(pd.DataFrame, pd.DataFrame)\n",
    "* generate_RS_presence_counts(R:pd.DataFrame, S:pd.DataFrame)->pd.DataFrame\n",
    "* compute_log_odds_ratio(RS_counts_df:pd.DataFrame)->pd.DataFrame\n",
    "* get_cluster_resistance_LOR(presence_df:pd.DataFrame, pheno_df:pd.DataFrame)-> pd.DataFrame\n",
    "    \n",
    "'''\n",
    "\n",
    "\n",
    "import requests\n",
    "import re\n",
    "import os\n",
    "import sys\n",
    "import networkx as nx\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def generate_files_paths(pipeline_output_path):\n",
    "    '''\n",
    "    takes the output dir path of the pipeline for a species, and returns the 4 file paths inside of it needed to extract the data\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    pipeline_output_path: str, path to the pipeline output directory\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    clstr_path: str, path to the cluster file, cdhit output\n",
    "    freq_path: str, path to the cluster frequency file\n",
    "    pan_annot_path: str, path to the pangenome annotation file\n",
    "    clstr_fasta_path: str, path to the cluster fasta file, aslo cd hit output\n",
    "    '''\n",
    "\n",
    "    pipeline_output_path=pipeline_output_path.rstrip('/')\n",
    "    basename=os.path.basename(pipeline_output_path)\n",
    "\n",
    "    clstr_path=pipeline_output_path+'/'+basename+'.fasta.clstr'\n",
    "    freq_path=pipeline_output_path+'/'+basename+'_cluster_frequencies.csv'\n",
    "    pan_annot_path=pipeline_output_path+'/'+basename+'_pangenome.csv'\n",
    "    clstr_fasta_path=pipeline_output_path+'/'+basename+'.fasta'\n",
    "\n",
    "    return clstr_path, freq_path, pan_annot_path, clstr_fasta_path\n",
    "\n",
    "# clstr_path, freq_path, pan_annot_path, clstr_fasta_path= generate_files_paths(\"../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/\")\n",
    "\n",
    "def get_cluster_frequency(freq_path):\n",
    "    '''\n",
    "    Takes the path to the cluster frequency file and returns a dataframe with the cluster frequencies\n",
    "\n",
    "    Parameters:\n",
    "    freq_path: str\n",
    "\n",
    "    Returns:\n",
    "    freq_df: pandas.DataFrame\n",
    "    '''\n",
    "\n",
    "    freq_df=pd.read_csv(freq_path, index_col=0)\n",
    "    freq_df.columns=['Cluster','frequency']\n",
    "    freq_df.index=freq_df['Cluster']\n",
    "    freq_df.drop('Cluster', axis=1, inplace=True)\n",
    "\n",
    "    return freq_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_representative_products(clstr_fasta_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a CD-HIT fasta output and returns a dataframe that matches each CDS PATRIC ID with its product name\n",
    "\n",
    "    the fasta output is of this form:\n",
    "\n",
    "    >fig|195.2024.peg.83 Putative oxidoreductase ferredoxin-type protein, clusters with CPO\n",
    "    MNFSQISDACVKCGKCIPVCTIHEVNRDETTSPRGFLDLLAAYKEEKLELDKEAKKIFES\n",
    "    CFLCTNCVEVCPSKLRVDNVIEEVRYDIAKKFGIAWYKKIIFFFLRRRKILDLVAKLGYV\n",
    "    FQSCAFKIQSQDQNVGMKAKFSMPFVKKGRLLTSFNKKSFLNSNPDFIDNGGEKTVGFFV\n",
    "    GCLANYFYIDTANAVLKIAKEVKINVDLMKEQVCCGAPQFFTGDFKSVEILAKKNIEYFE\n",
    "    KKLEKLDAIIIPEATCSAMLKIDLEHFFNMQNEPEWAKRAQKISSRIYMASEYFYKFTNL\n",
    "    KELLESKKKLNYSITYHDPCHARKMQGVFKEPRELLKANYHFVEMSNPNACCGFGGVSMQ\n",
    "    TDYYDRALSVGLKKASMIDESKACVVSAECSACRMQISNALEQNSSKAIFASPLELIAKA\n",
    "    L\n",
    "    >fig|195.2024.peg.542 Septum-associated rare lipoprotein A\n",
    "    MKPYTINGKTYYPTVVSVGETADGIASWYGPGFHGKKTSNGETYNQNGLTAAHKTLPMNT\n",
    "    ILKVTNLNNNRQVTVRVNDRGPFVNNRIIDLSKGAASQIDMIAAGTAPVRLEVIGFGSAN\n",
    "    SGNNVVHSNINYGASGGIANNGQIYEGGNFMVQIGAFKNPSGAQTIASRYKTYRTYSSTI\n",
    "    RKSSVDGLSRVFLTGFRSEEEARDFAASGAFAGAFVVRE\n",
    "\n",
    "    The aim is to match 195.2024.peg.542 with Septum-associated rare lipoprotein A in teh dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_fasta_path: str, path of the fasta output\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, df of columns gene_representative and product_name\n",
    "    '''\n",
    "\n",
    "    dict={}\n",
    "\n",
    "    with open(clstr_fasta_path,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith(\">\"):\n",
    "                pattern=\">fig.(\\d+\\.\\d+.peg.\\d+) (.+)\"\n",
    "                gene_representative=re.match(string=line, pattern=pattern).group(1)\n",
    "                product_name=re.match(string=line, pattern=pattern).group(2)\n",
    "                dict[gene_representative]=product_name\n",
    "    \n",
    "\n",
    "    df = pd.DataFrame(dict.items(), columns=['gene_representative', 'product_name'])\n",
    "    \n",
    "    return df\n",
    "\n",
    "def combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Combines the cluster representative dataframe with the product name dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_rep_df: pd.DataFrame, dataframe with cluster number and gene representative (get_cluster_representatives() output)\n",
    "    - rep_prod_df: pd.DataFrame, dataframe with gene representative and product name (get_representative_products() output)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, dataframe with cluster number, gene representative and product name\n",
    "    '''\n",
    "    # if clstr rep df doesnt have a column named Cluster, create one out of index\n",
    "    if 'Cluster' not in clstr_rep_df.columns:\n",
    "        clstr_rep_df['Cluster']=clstr_rep_df.index\n",
    "\n",
    "    df=pd.merge(clstr_rep_df, rep_prod_df, on='gene_representative', how='left')\n",
    "    df.drop(columns='gene_representative', inplace=True)\n",
    "\n",
    "    df.set_index('Cluster', inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a pangenome annotation file and returns a dataframe with cluster number and gene class\n",
    "    This annotation file is a <species>_pangenome.csv file output from the pangenome analysis pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - pan_annot_path: str, path to the pangenome annotation file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - pan_df: pd.DataFrame, dataframe with cluster number and gene class\n",
    "    '''\n",
    "    with open(pan_annot_path, 'r') as f:\n",
    "        pan_df=pd.read_csv(f, index_col=0)\n",
    "        pan_df.drop(pan_df.columns[[1]], axis=1, inplace=True)\n",
    "        pan_df.columns=['Cluster', 'pan_gene_class']\n",
    "\n",
    "    pan_df.set_index('Cluster', inplace=True)\n",
    "    return pan_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path):\n",
    "\n",
    "    clstr_patric_id_df=get_cluster_representatives(clstr_path)\n",
    "    clstr_gene_class_df= get_cluster_pan_gene_class(pan_annot_path)\n",
    "    _patric_id_product_df=get_representative_products(clstr_fasta_path)\n",
    "    clstr_product_df=combine_cluster_product(clstr_patric_id_df, _patric_id_product_df)\n",
    "    clstr_freq_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "    df=pd.concat([clstr_patric_id_df, clstr_gene_class_df, clstr_product_df, clstr_freq_df], axis=1)\n",
    "\n",
    "    #if there exist a Cluster col remove\n",
    "    if 'Cluster' in df.columns:\n",
    "        df.drop('Cluster', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def create_dicts(df:pd.DataFrame):\n",
    "    '''\n",
    "    takes a nxm data frame and returns m dictionaries of keys the index, and values each col repectively\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - dicts: dict, dict of m dictionaries - key is the col name, value is the dictionary of the col\n",
    "    '''\n",
    "    dicts = {col: df[col].to_dict() for col in df.columns}\n",
    "    return dicts\n",
    "\n",
    "def set_node_attributes_by_cluster(df:pd.DataFrame, G:nx.graph):\n",
    "    '''\n",
    "    takes a dataframe and a graph and sets the node attributes for each cluster\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "    - G: nx.Graph, graph to set the node attributes\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - G: nx.Graph, graph with node attributes set\n",
    "    '''\n",
    "    dicts = create_dicts(df)\n",
    "\n",
    "    for key in dicts.keys():\n",
    "        nx.set_node_attributes(G, dicts[key], key)\n",
    "\n",
    "    return G\n",
    "\n",
    "_output_path='../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/'\n",
    "\n",
    "def set_output_path(path:str):\n",
    "    '''\n",
    "    sets the output path for the pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - path: str, path to the output directory\n",
    "    '''\n",
    "    global _output_path\n",
    "    _output_path=path\n",
    "    return _output_path\n",
    "\n",
    "def set_cluster_attributes(G:nx.Graph, pipeline_output=_output_path)->nx.Graph:\n",
    "    '''\n",
    "    takes a Graph and the pipeline's output path and performs all of the following: \n",
    "\n",
    "    - get all the files paths required for the attributes\n",
    "    - generate df for each and concatenate them all in one of index Cluster\n",
    "    - create dictionaries out of it and perform nx.set_node_attributes using each on G\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    G: nx.Graph, network of nodes Clusters\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    G: nx.Graph, network of nodes Clusters, attributes: pan_gene_class, gene_representative, pan_gene_class, frequency\n",
    "    '''\n",
    "\n",
    "    p1, p2, p3, p4 = generate_files_paths(pipeline_output)\n",
    "    df = get_cluster_attributes(p1, p2, p3, p4)\n",
    "    G = set_node_attributes_by_cluster(df, G)\n",
    "\n",
    "    return G\n",
    "\n",
    "# --------------- LOR calculation ----------------\n",
    "\n",
    "def generate_phenotype_df(pheno_path:str, presence_df:pd.DataFrame=None)->pd.DataFrame:\n",
    "    '''\n",
    "    takes in a path to a phenotype file to make a df out of it\n",
    "    if a presence_df is provided, it will filter the pheno_df to only include samples that are in the presence_df\n",
    "\n",
    "    param:\n",
    "    ----\n",
    "        - pheno_path: str\n",
    "        - presence_df: pd.DataFrame (optional)\n",
    "\n",
    "    return:\n",
    "    ----\n",
    "        - pheno_df: pd.DataFrame\n",
    "    '''\n",
    "    pheno_df = pd.read_csv(pheno_path, index_col=0)\n",
    "\n",
    "    # print('pheno dim:', pheno_df.shape)\n",
    "    \n",
    "    # -- convert presence_df cols to str temporarily to trace error\n",
    "    presence_df.columns = presence_df.columns.astype(str)\n",
    "\n",
    "    #samples stripped\n",
    "    if presence_df is not None:\n",
    "        samples_presence= presence_df.columns\n",
    "        mask = pheno_df.index.astype(str).isin(samples_presence)\n",
    "        # print(mask)\n",
    "        # print(pheno_df.index)\n",
    "        # print(samples_presence) #same type\n",
    "        pheno_df = pheno_df.loc[mask]\n",
    "\n",
    "    # print('pheno dim:', pheno_df.shape)\n",
    "\n",
    "    return pheno_df\n",
    "\n",
    "def split_matrix_by_phenotype(unlabeled_presence_df:pd.DataFrame, pheno_df: pd.DataFrame)->(pd.DataFrame, pd.DataFrame):\n",
    "    '''\n",
    "    takes a gene absence presence dataframe (n x m) with the samples classification and splits it to 2 dataframes:\n",
    "        - one that represents all R samples (samples found in R_list)\n",
    "        - one that represents all S samples (found in S_list)\n",
    "    and returns them in this order: R then S\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "        - unlabeled_presence_df: (pd.DataFrame) the data frame (GxS)\n",
    "        - R_list: (list) samples genome_id that have a R (1) phenotype\n",
    "        - S_list: (list) samples having the 0 phenotype\n",
    "    \n",
    "    return:\n",
    "    ------\n",
    "        - R_df: (pd.DataFrame) data frame where columns are only for R samples\n",
    "        - S_df: (pd.DataFrame) df where cols are only for S samples\n",
    "\n",
    "    '''\n",
    "\n",
    "\n",
    "    # make sure pheno_df doesnthave more samples than those specified in the presence_df, subsample it otherwise\n",
    "    samples_presence= unlabeled_presence_df.columns\n",
    "    mask = pheno_df.index.astype(str).isin(samples_presence)\n",
    "    pheno_df = pheno_df.loc[mask]\n",
    "\n",
    "\n",
    "    # get the list of R and S samples:\n",
    "    R=[];S=[];U=[]\n",
    "    for sample in pheno_df.index:\n",
    "\n",
    "        if pheno_df.loc[sample].values[0]==1:\n",
    "            sample=str(sample)\n",
    "            R.append(str(sample))\n",
    "        elif pheno_df.loc[sample].values[0]==0:\n",
    "            sample=str(sample)\n",
    "            S.append(str(sample))\n",
    "        else:\n",
    "            sample=str(sample)\n",
    "            U.append(str(sample))\n",
    "\n",
    "    # All we care for care R and S which designate the list of resistant and susceptible samples respectively\n",
    "\n",
    "    R_df = unlabeled_presence_df[R]\n",
    "    S_df = unlabeled_presence_df[S]\n",
    "\n",
    "\n",
    "\n",
    "    return R_df, S_df\n",
    "\n",
    "def generate_RS_presence_counts(R:pd.DataFrame, S:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes the presence matrix of R samples and S samples and returns a dataframe of the count of genes present and absent in each group of samples.\n",
    "\n",
    "    *This will memoize the entries needed for the log odds ratio computation.*\n",
    "\n",
    "    \n",
    "    e.g., output (w\\o the log odds):\n",
    "\n",
    "    | Gene       | R_present | R_absent | S_present | S_absent |\n",
    "    |------------|-----------|----------|-----------|----------|\n",
    "    | group_1001 | 96        | 0        | 187       | 0        |\n",
    "    | tig        | 96        | 0        | 187       | 0        |\n",
    "    | legF_1     | 96        | 0        | 187       | 0        |\n",
    "\n",
    "    param:\n",
    "    ----------------\n",
    "        - R: (pd.DataFrame) presence matrix of R samples (output of get_subdf_cols)\n",
    "        - S: (pd.DataFrame) presence matrix of S samples\n",
    "\n",
    "    return:\n",
    "    ----------------\n",
    "        - new_df: (pd.DataFrame) dataframe of the count of genes present and absent in each group of samples.\n",
    "\n",
    "    NOTE: it will perform the 0.5 correction - If for one gene any of the counts is 0, it will add 0.5 to all counts for that gene.\n",
    "\n",
    "    '''\n",
    "    R_present=R.sum(axis=1)\n",
    "    R_absent=R.shape[1] - R_present\n",
    "    S_present=S.sum(axis=1)\n",
    "    S_absent=S.shape[1]-S_present\n",
    "\n",
    "    indices=R.index\n",
    "\n",
    "    new_df=pd.DataFrame({'R_present':R_present, 'R_absent':R_absent, 'S_present':S_present, 'S_absent':S_absent})\n",
    "    new_df.index=indices\n",
    "\n",
    "    #the 0.5 correction:\n",
    "    for index in new_df.index:\n",
    "        row=list(new_df.loc[index])\n",
    "        if 0 in row: #check if any of the cols have value 0\n",
    "            new_df.loc[index]=new_df.loc[index]+0.5\n",
    "\n",
    "    return new_df\n",
    "\n",
    "def compute_log_odds_ratio(RS_counts_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    takes the RS_counts_df and computes the log odds ratio for each gene in the dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "        - RS_counts_df: pd.DataFrame\n",
    "\n",
    "\n",
    "    return:\n",
    "    ------\n",
    "        - log_odds_df: pd.DataFrame\n",
    "    '''\n",
    "\n",
    "    # n_R=len(R); n_S=len(S)\n",
    "\n",
    "\n",
    "    R_present=RS_counts_df['R_present']\n",
    "    R_absent=RS_counts_df['R_absent']\n",
    "    S_present=RS_counts_df['S_present']\n",
    "    S_absent=RS_counts_df['S_absent']\n",
    "\n",
    "    log_odds=np.log((R_present/R_absent)/(S_present/S_absent))\n",
    "\n",
    "    # print(log_odds)\n",
    "\n",
    "    df=pd.DataFrame({\"log_odds\":log_odds})\n",
    "    return df\n",
    "\n",
    "def get_cluster_resistance_LOR(presence_df:pd.DataFrame, pheno_path:str)-> pd.DataFrame:\n",
    "    '''\n",
    "    Knits all the steps from splitting the matrix to computing the log odds ratio for each gene in the matrix\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "        * presence_df: pd.DataFrame, GxS\n",
    "        * pheno_df: str, path to the phenotype file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "        * log_odds_df: pd.DataFrame, Gx1\n",
    "    '''\n",
    "    pheno_df = generate_phenotype_df(pheno_path, presence_df)\n",
    "    # print(pheno_df)\n",
    "    \n",
    "    R,S=split_matrix_by_phenotype(presence_df, pheno_df)\n",
    "    RS_counts = generate_RS_presence_counts(R, S)\n",
    "    log_odds_df = compute_log_odds_ratio(RS_counts)\n",
    "\n",
    "    return log_odds_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rayane/miniconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: invalid value encountered in log\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "# -- this works\n",
    "\n",
    "pheno = '../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv'\n",
    "presence=pd.read_csv('../../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv', index_col=0 )\n",
    "\n",
    "lor_df= get_cluster_resistance_LOR(pheno_path=pheno, presence_df=presence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['195.2024', '195.2025', '195.2026', '195.2027', '195.2028', '195.2029',\n",
       "       '195.203', '195.2031', '195.2032', '195.2033',\n",
       "       ...\n",
       "       '195.2307', '195.2308', '195.231', '195.2311', '195.2312', '195.2313',\n",
       "       '195.2314', '195.2315', '195.2316', '195.2317'],\n",
       "      dtype='object', length=269)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Float64Index([195.2047, 195.2058, 195.2057, 195.2056, 195.2055, 195.2052,\n",
       "              195.2053, 195.2051, 195.2049, 195.2059,\n",
       "              ...\n",
       "              195.2164, 195.2165, 195.2166, 195.2169,  195.217, 195.2171,\n",
       "              195.2172, 195.2159, 195.2178, 195.2298],\n",
       "             dtype='float64', name='genome_id', length=283)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Float64Index([195.2024, 195.2025, 195.2026, 195.2027, 195.2028, 195.2029,\n",
       "               195.203, 195.2031, 195.2032, 195.2033,\n",
       "              ...\n",
       "              195.2307, 195.2308,  195.231, 195.2311, 195.2312, 195.2313,\n",
       "              195.2314, 195.2315, 195.2316, 195.2317],\n",
       "             dtype='float64', length=269)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "species='Campylobacter_coli'; drug='ciprofloxacin'\n",
    "\n",
    "# ------------------ FIX ------------\n",
    "# exceptionally this way to C coli (presence matrix wasnt standardized to GxS)\n",
    "X_df = pd.read_csv('../../data/presence_matrices/Campylobacter_coli_presence_absence.csv', index_col=0)\n",
    "presence_df=X_df.T\n",
    "\n",
    "y_df= pd.read_csv('../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv', index_col=0)\n",
    "pheno_df=y_df\n",
    "\n",
    "for gene in X_df.columns:\n",
    "    \n",
    "    if X_df[gene].std() == 0:\n",
    "        X_df.drop(gene, axis=1, inplace=True)\n",
    "\n",
    "X_df = X_df.sort_index()\n",
    "y_df = y_df.sort_index()\n",
    "\n",
    "y_indices=list(y_df.index)\n",
    "X_indices=list(X_df.index)\n",
    "\n",
    "intersection = [i for i in y_indices if i in X_indices]\n",
    "y_df = y_df.loc[intersection]\n",
    "X_df = X_df.loc[intersection]\n",
    "\n",
    "X = X_df.values\n",
    "y = y_df.values\n",
    "\n",
    "labeled_matrix = pd.concat([X_df, y_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pheno dim: (283, 1)\n",
      "[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True False False False False\n",
      " False False False False False  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      " False  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True False  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True False  True  True\n",
      "  True  True  True  True  True  True False False False  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True False  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SIR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>195.2047</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2058</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2057</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2056</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2055</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2171</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2172</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2159</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2178</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195.2298</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           SIR\n",
       "genome_id     \n",
       "195.2047     1\n",
       "195.2058     1\n",
       "195.2057     0\n",
       "195.2056     0\n",
       "195.2055     0\n",
       "...        ...\n",
       "195.2171     0\n",
       "195.2172     1\n",
       "195.2159     1\n",
       "195.2178     0\n",
       "195.2298     1\n",
       "\n",
       "[267 rows x 1 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_phenotype_df('../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv', presence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pheno dim: (283, 1)\n",
      "[ True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True False False False False\n",
      " False False False False False  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      " False  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True False  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True False  True  True\n",
      "  True  True  True  True  True  True False False False  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True False  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True]\n",
      "           SIR\n",
      "genome_id     \n",
      "195.2047     1\n",
      "195.2058     1\n",
      "195.2057     0\n",
      "195.2056     0\n",
      "195.2055     0\n",
      "...        ...\n",
      "195.2171     0\n",
      "195.2172     1\n",
      "195.2159     1\n",
      "195.2178     0\n",
      "195.2298     1\n",
      "\n",
      "[267 rows x 1 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rayane/miniconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: invalid value encountered in log\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log_odds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>1.270365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>-1.220273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>0.104140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>-0.311244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              log_odds\n",
       "Cluster 0     1.270365\n",
       "Cluster 1    -1.220273\n",
       "Cluster 2     0.104140\n",
       "Cluster 3    -0.311244\n",
       "Cluster 4     0.311244\n",
       "...                ...\n",
       "Cluster 8016 -0.311244\n",
       "Cluster 8017 -0.311244\n",
       "Cluster 8018 -0.311244\n",
       "Cluster 8019 -0.311244\n",
       "Cluster 8020 -0.311244\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_odds_df = get_cluster_resistance_LOR(presence_df=presence_df, pheno_path='../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv')\n",
    "log_odds_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
